File: src/main/java/toby/command/commands/misc/EightBallCommand.java
package toby.command.commands.misc;

import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent;
import toby.command.CommandContext;
import toby.command.ICommand;
import toby.jpa.dto.UserDto;
import toby.jpa.service.IUserService;

import java.util.Random;

public class EightBallCommand implements IMiscCommand {


    public static final long TOMS_DISCORD_ID = 312691905030782977L;
    private final IUserService userService;

    public EightBallCommand(IUserService userService) {

        this.userService = userService;
    }

    @Override
    public void handle(CommandContext ctx, UserDto requestingUserDto, Integer deleteDelay) {
        SlashCommandInteractionEvent event = ctx.getEvent();
        event.deferReply().queue();
        ICommand.deleteAfter(event.getHook(), deleteDelay);

        Random r = new Random();

        int choice = 1 + r.nextInt(20);
        String response = switch (choice) {
            case 1 -> "It is certain";
            case 2 -> "It is decidedly so";
            case 3 -> "Without a doubt";
            case 4 -> "Yes - definitely";
            case 5 -> "You may rely on it";
            case 6 -> "As I see it, yes";
            case 7 -> "Most likely";
            case 8 -> "Outlook good";
            case 9 -> "Signs point to yes";
            case 10 -> "Yes";
            case 11 -> "Reply hazy, try again";
            case 12 -> "Ask again later";
            case 13 -> "Better not tell you now";
            case 14 -> "Cannot predict now";
            case 15 -> "Concentrate and ask again";
            case 16 -> "Don't count on it";
            case 17 -> "My reply is no";
            case 18 -> "My sources say no";
            case 19 -> "Outlook not so good";
            case 20 -> "Very doubtful";
            default -> "I fucked up, please try again";
        };

        if(requestingUserDto.getDiscordId().equals(TOMS_DISCORD_ID)){
            event.replyFormat("MAGIC 8-BALL SAYS: Don't fucking talk to me.").queue(message -> ICommand.deleteAfter(message, deleteDelay));
            Long socialCredit = requestingUserDto.getSocialCredit();
            int deductedSocialCredit = -5 * choice;
            requestingUserDto.setSocialCredit(socialCredit + deductedSocialCredit);
            event.replyFormat("Deducted: %d social credit.", deductedSocialCredit).queue(message -> ICommand.deleteAfter(message, deleteDelay));
            userService.updateUser(requestingUserDto);
            return;
        }
        event.replyFormat("MAGIC 8-BALL SAYS: %s.", response).queue(message -> ICommand.deleteAfter(message, deleteDelay));
    }

    @Override
    public String getName() {
        return "8ball";
    }

    @Override
    public String getDescription() {
        return "Think of a question and let me divine to you an answer!";
    }

}

File: discord-bot/src/main/kotlin/bot/toby/command/commands/misc/EightBallCommand.kt
package bot.toby.command.commands.misc

import core.command.CommandContext
import database.dto.user.UserDto
import database.service.user.UserService
import net.dv8tion.jda.api.interactions.InteractionHook
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component
import java.util.Random
import java.util.concurrent.TimeUnit

@Component
class EightBallCommand @Autowired constructor(private val userService: UserService) : MiscCommand {

    override fun handle(ctx: CommandContext, requestingUserDto: UserDto, deleteDelay: Int) {
        val event = ctx.event
        ask(event.hook, requestingUserDto, event.user.effectiveName, deleteDelay)
    }

    /**
     * Shared entry point so the `EightBallButton` re-ask flow renders the
     * same two-stage shake→reveal as a fresh `/8ball` invocation. Caller
     * is responsible for acknowledging the interaction first
     * (`deferReply()` for slash, `deferEdit()` for the button) so the
     * hook is ready to edit the original message.
     */
    fun ask(hook: InteractionHook, requestingUserDto: UserDto, askedBy: String, deleteDelay: Int) {
        val choice = 1 + Random().nextInt(20)
        val response = RESPONSES[choice - 1]

        if (requestingUserDto.discordId == TOMS_DISCORD_ID) {
            val deducted = -5 * choice
            requestingUserDto.socialCredit = (requestingUserDto.socialCredit ?: 0L) + deducted
            userService.updateUser(requestingUserDto)
            hook.editOriginalEmbeds(listOf(EightBallEmbeds.tomEmbed(deducted))).queue {
                if (deleteDelay > 0) it.delete().queueAfter(deleteDelay.toLong(), TimeUnit.SECONDS)
            }
            return
        }

        // Collection overload of editOriginalEmbeds rather than the
        // varargs one — mockk's vararg matchers don't play nicely with
        // the JDA default-method dispatch, and this is exactly one
        // embed each call anyway.
        hook.editOriginalEmbeds(listOf(EightBallEmbeds.shakeEmbed())).queue()
        hook.editOriginalEmbeds(listOf(EightBallEmbeds.answerEmbed(response, askedBy)))
            .setComponents(EightBallEmbeds.askAgainRow())
            .queueAfter(REVEAL_DELAY_MS, TimeUnit.MILLISECONDS) {
                if (deleteDelay > 0) it.delete().queueAfter(deleteDelay.toLong(), TimeUnit.SECONDS)
            }
    }

    override val name: String get() = "8ball"
    override val description: String get() = "Think of a question and let me divine to you an answer!"

    companion object {
        // A Discord snowflake, not a credential. The shape trips Qodana's
        // "discord-client-id" hardcoded-password heuristic.
        @Suppress("HardcodedPasswords")
        const val TOMS_DISCORD_ID = 312691905030782977L

        const val REVEAL_DELAY_MS = 1200L

        val RESPONSES = listOf(
            "It is certain",
            "It is decidedly so",
            "Without a doubt",
            "Yes - definitely",
            "You may rely on it",
            "As I see it, yes",
            "Most likely",
            "Outlook good",
            "Signs point to yes",
            "Yes",
            "Reply hazy, try again",
            "Ask again later",
            "Better not tell you now",
            "Cannot predict now",
            "Concentrate and ask again",
            "Don't count on it",
            "My reply is no",
            "My sources say no",
            "Outlook not so good",
            "Very doubtful",
        )
    }
}
